fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvPartitionCompensator.java

```diff
@@ -0,0 +1,696 @@
+            Column partitionColumn,
+            MaterializedView mv,
+            Set<String> mvPartitionNamesToRefresh) throws AnalysisException {
+        Set<String> refBaseTableUpdatedPartitionNames = mv.getUpdatedPartitionNamesOfTable(partitionByTable, true);
+        List<Range<PartitionKey>> refBaseTableRanges = getLatestPartitionRange(partitionByTable, partitionColumn,
+                refBaseTableUpdatedPartitionNames, MaterializedView.getPartitionExpr(mv));
+        // date to varchar range
+        Map<Range<PartitionKey>, Range<PartitionKey>> baseRangeMapping = null;
+        boolean isConvertToDate = PartitionUtil.isConvertToDate(mv.getFirstPartitionRefTableExpr(), partitionColumn);
+        if (isConvertToDate) {
+            baseRangeMapping = Maps.newHashMap();
+            // convert varchar range to date range
+            List<Range<PartitionKey>> baseTableDateRanges = Lists.newArrayList();
+            for (Range<PartitionKey> range : refBaseTableRanges) {
+                Range<PartitionKey> datePartitionRange = convertToDateRange(range);
+                baseTableDateRanges.add(datePartitionRange);
+                baseRangeMapping.put(datePartitionRange, range);
+            }
+            refBaseTableRanges = baseTableDateRanges;
+        }
+        // materialized view latest partition ranges except to-refresh partitions
+        List<Range<PartitionKey>> mvRanges = getLatestPartitionRangeForNativeTable(mv, mvPartitionNamesToRefresh);
+
+        List<Range<PartitionKey>> latestBaseTableRanges = Lists.newArrayList();
+        for (Range<PartitionKey> range : refBaseTableRanges) {
+            // if materialized view's partition range can enclose the ref base table range, we think that
+            // the materialized view's partition has been refreshed and should be compensated into the materialized
+            // view's partition predicate.
+            if (mvRanges.stream().anyMatch(mvRange -> mvRange.encloses(range))) {
+                latestBaseTableRanges.add(range);
+            }
+        }
+        if (isConvertToDate) {
+            // treat string type partition as list, so no need merge
+            List<Range<PartitionKey>> tmpRangeList = Lists.newArrayList();
+            for (Range<PartitionKey> range : latestBaseTableRanges) {
+                tmpRangeList.add(baseRangeMapping.get(range));
+            }
+            return tmpRangeList;
+        } else {
+            return MvUtils.mergeRanges(latestBaseTableRanges);
+        }
+    }
+
+    private static List<Range<PartitionKey>> getLatestPartitionRangeForNativeTable(OlapTable partitionTable,
+                                                                                   Set<String> modifiedPartitionNames) {
+        // partitions that will be excluded
+        Set<Long> filteredIds = Sets.newHashSet();
+        for (Partition p : partitionTable.getPartitions()) {
+            if (modifiedPartitionNames.contains(p.getName()) || !p.hasData()) {
```

Comment:
`hasData` is not always correct, what if it returns staled result ?

Reply:
no matter? This will filter partitions which are contained by mvPartitionNamesToRefresh and partitions have no data which no need for partition compensate.